File: count_and_say.cpp
#include<iostream>
#include<vector>
#include<string>
#include<unordered_map>
using namespace std;

    vector<string> spliter(string x){
        vector<string> return_string;
        int head = 0, i = 0;
        for(i=0; i<x.length()-1; i++){
            if(x[i] != x[i+1]){
                return_string.push_back(x.substr(head, i+1-head));
                head = i+1;
            }
        }
        return_string.push_back(x.substr(head, x.length()-head));
        return return_string;
    }

    string talker(string x){
        string talk = "";
        for(string ele : spliter(x))
            talk += to_string(ele.length())+ele[0];
        return talk;
    }

    unordered_map<string, string> history;
    unordered_map<string, string>::iterator iter;
    string countAndSay(int n) {
        string start = "1";
        for(int i=1; i<n; i++){
            iter = history.find(to_string(i));
            if(iter != history.end()){
                start = iter->second;
            }else{
                start = talker(start);
                history[to_string(i)] = start;
            }
        }
        return start;
    }


int main(){
    cout<<countAndSay(5)<<endl;
}

File: python3/count_and_say.py
def spliter(x):
    x = str(x)
    return_string = []
    head = 0
    i = 0
    for i in range(len(x)-1):
        if x[i] != x[i+1]:
            return_string.append(x[head:i+1])
            head = i+1
    return_string.append(x[head:])
    return return_string

def talker(x):
    talk = ''
    for ele in spliter(x):
        talk += str(len(ele))+ele[0]
    return talk

history = dict()
def countAndSay(n):
    start = 1
    for i in range(1, n):
        if i in history:
            start = history[i]
        else:
            start = talker(start)
            history[i] = start
    return str(start)


print(countAndSay(5))

